### DISCUSSION TITLE
Computed properties based on object state/input?

### DISCUSSION THREAD

**Discussion by ovp87** [spatie/laravel-data#261] [Q&A] [answered]
Given the following example

```
<?php declare(strict_types = 1);

namespace App;

use Spatie\LaravelData\Data;
use Spatie\LaravelData\Lazy;

class Test extends Data
{
    public Lazy|bool $test1;
    public Lazy|int $test2;

    public function __construct(public string $name)
    {
        $this->test1 = Lazy::create(function(){
            return $this->name === 'hello world';
        });

        $this->test2 = Lazy::create(function(){
            return $this->name === 'hello world' 
                ? 1234 
                : 5678;
        });
    }
}
```
`
$test = \App\Test::validateAndCreate(['name' => 'hello world']);`

`$test2` fails validation (evaluated as required), while `$test1` is set correctly. 

I'm assuming i'm using it incorrectly, what i'm really trying to do is automatically set a property on the object, based on the input/state of the object, achieving something akin to computed properties in vue. 
Ideally without it being possible to overwrite this property by passing in ` 'test1' => false`.

Can anyone give me a pointer to the best way of doing this for this package?

**Comment by ovp87** [comment, marked as answer]
"Closest" i've come is something like this:

```
class Test extends Data
{
    #[Nullable]
    public readonly bool $test1;
    #[Nullable]
    public readonly int $test2;

    public function __construct(public string $name) {
        $this->test1 = $name === 'hello world';
        $this->test2 = $name === 'hello world' ? 1234 : 5678;
    }

}
```

**Comment by Nir-An** [comment]
Hi @ovp87 

What about native '?' without  `` #[Nullable]``

```php


class Test extends Data
{
    public readonly ?bool $test1;

    public readonly ?int $test2;

    public function __construct(public string $name) {
        $this->test1 = $name === 'hello world';
        $this->test2 = $name === 'hello world' ? 1234 : 5678;
    }

}


```

https://wiki.php.net/rfc/typed_properties_v2

**Reply by ovp87** [reply]
That would be "cleaner", however i run into `Cannot modify readonly property` issues when using readonly properties. 
It's not clear to me how `?bool` differs from using the nullable attribute.

**Reply by ejntaylor** [reply]
```    public function __construct()
public function __construct()
    {
        [$projectName, $manifestFile] = explode(':', "{$this->name}:");
        $this->projectName = $projectName;
        $this->manifestFile = $manifestFile;
    }
 ```
 
 Wondering how this is possible when using an existing parameter. Currently getting this Typed Property Exception:
 
 `\Project::$name must not be accessed before initialization`

**Comment by rubenvanassche** [comment]
Have you already taken a look at optional properties? https://spatie.be/docs/laravel-data/v2/as-a-data-transfer-object/optional-properties

**Comment by XedinUnknown** [comment]
@rubenvanassche, it says 

> The value of artist will automatically be set to `Optional`. When you transform this data object to an array, it will look like this:
>
> ```php
> [
>     'title' => 'Never gonna give you up'
> ]
> ```

But it doesn't say how to determine the value of `Optional`, and I don't understand how it becomes `'Never gonna give you up'`.

Either way, that isn't the point for me. In my case, as a diligent developer, I have an `enum` that lists the possible values that the `type` property of my entity may have, and is used throughout the application. My OpenAPI specification uses a corresponding `enum` schema in order to allow keys from that enum, only those, and at least one:

```yaml
type: object
minProperties: 1
propertyNames:
  $ref: MyEnum
additionalProperties:
  type: string
```

This is great, because in order to allow anothe property (all have the same type), I only need to add a new value to `MyEnum`.
How do I implement this using a DTO?

I can use an associative array e.g.

```
/* @var array<value-of<MyEnum>, string> */
public readonly array $names;
```

But now my DTO translates to having a property `names`, rather than having a bunch of properties - each one of my enum values.

Right now, I have to resort to adding a separate property to the DTO, one for each enum member.
